src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/CodeActions/CodeActionsHandler.ts

```diff
@@ -0,0 +1,71 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { RequestType } from 'vscode-languageclient';
+import { RazorDocumentManager } from '../RazorDocumentManager';
+import { RazorDocumentSynchronizer } from '../RazorDocumentSynchronizer';
+import { RazorLanguageFeatureBase } from '../RazorLanguageFeatureBase';
+import { RazorLanguageServerClient } from '../RazorLanguageServerClient';
+import { RazorLanguageServiceClient } from '../RazorLanguageServiceClient';
+import { RazorLogger } from '../RazorLogger';
+import { RazorCodeAction } from '../RPC/RazorCodeAction';
+import { SerializableCodeActionParams } from '../RPC/SerializableCodeActionParams';
+import { convertRangeFromSerializable } from '../RPC/SerializableRange';
+
+export class CodeActionsHandler
+    extends RazorLanguageFeatureBase {
+
+    private static readonly getCodeActionsEndpoint = 'razor/provideCodeActions';
+    private codeActionRequestType: RequestType<SerializableCodeActionParams, RazorCodeAction[], any, any> = new RequestType(CodeActionsHandler.getCodeActionsEndpoint);
+    private emptyCodeActionResponse: RazorCodeAction[] = [];
+
+    constructor(
+        documentSynchronizer: RazorDocumentSynchronizer,
+        documentManager: RazorDocumentManager,
+        serviceClient: RazorLanguageServiceClient,
+        private readonly serverClient: RazorLanguageServerClient,
+        logger: RazorLogger) {
+            super(documentSynchronizer, documentManager, serviceClient, logger);
+    }
+
+    public register() {
+        // tslint:disable-next-line: no-floating-promises
+        this.serverClient.onRequestWithParams<SerializableCodeActionParams, RazorCodeAction[], any, any>(
+            this.codeActionRequestType,
+            async (request, token) => this.getCodeActions(request, token));
+    }
+
+    private async getCodeActions(
+        codeActionParams: SerializableCodeActionParams,
+        token: vscode.CancellationToken) {
+        try {
+            const razorDocumentUri = vscode.Uri.parse(codeActionParams.textDocument.uri);
+            const razorDocument = await this.documentManager.getDocument(razorDocumentUri);
+            const virtualCsharpUri = razorDocument.csharpDocument.uri;
```

Comment:
Should validate that the `razorDocument` is available before doing this. If it was closed or deleted it'll be `null` here.

Reply:
Added `undefined` check.